Pipe improvement

samytichadou opened the pull request:
Turned the pipe getter into a method to be able to reuse it.
So now operators will check for pipe before the main execution, and try to update it (if audacity has been opened since blender launched)

In the addon prefs, there is now a property to point on the audacity executable to allow the addon to automatically launch it.
A downside is to have a waiting time, between calling audacity and having it open and ready. This waiting time, if too short, can cause error. The default seems ok (5s) but it can be set by user in prefs if needed.
An improvement could be to have an operator processing this automatically, but for now, seems ok.

Here is a little demo video : 
https://youtu.be/o2HQz504-ls

tin2tin commented:
Again, it just amazing how much you can do within a very small time span. Wow. 

Setting up and opening Audacity works well here. Later on, maybe we can see if Audacity can return more specific error messages. Maybe we can catch an invalid path on the Blender side though? 

It seems like the play function is suffering from the time spend on checks, when it plays it doesn't seems to be in sync, and stopping, doesn't stop Audacity. Maybe the play function should just have a "try" check instead of a fully fledged pipe check? 

I haven't looked at the code yet, but thank you so much for all the time you already have invested in this. :-)

samytichadou commented:
> 
> Setting up and opening Audacity works well here. Later on, maybe we can see if Audacity can return more specific error messages. Maybe we can catch an invalid path on the Blender side though?

For now, using what you collected for the pipe system, i can't really catch an error (or didn't get how to do it) but yeah, that would be a great improvement to keep in mind for the future ! And about an invalid path, you mean for the audacity exec ? it already checks for that, If the path is not a proper file, it return a warning message
> 
> It seems like the play function is suffering from the time spend on checks, when it plays it doesn't seems to be in sync, and stopping, doesn't stop Audacity. Maybe the play function should just have a "try" check instead of a fully fledged pipe check?

Yes that was one of my concerns, but it didn't seem to slow things down for me. I'll do what you suggested and just put a simple `try` in the play operator
> 
> I haven't looked at the code yet, but thank you so much for all the time you already have invested in this. :-)

Thanks for the kind words ! yeah, i have some spare time right now, and this pipe addon has some great potential ! happy to help here :)

samytichadou commented:
I changed a few things in order to fix the hanging when setting up the pipe.
To me, all the operators communicating with Audacity should have the same behavior, so i added a simple line in the gui :

![image](https://user-images.githubusercontent.com/26276003/110783902-4c681a80-8269-11eb-8c78-e7447160043b.png)

The pipe is check on every startup through the handler, if not existent, the communicating operators will not be available : 

![image](https://user-images.githubusercontent.com/26276003/110784035-74577e00-8269-11eb-9bd7-7f6ed4e0a832.png)

The user would have to hit the refresh button for the pipe, which will try to setting it up and launch Audacity. 

Then when using the operators, they will do a `try:` with a simple audacity command to test the pipe, if not available, warning message and abort of the execution. The ui will reflect it ("'Pipe not available")

Seems way more consistent and understandable to me, and it allows to wait a lot less for audacity to load, so i changed the default in the prefs.

[Here is a little demo video](https://youtu.be/cyLcfL7Crxw)

tin2tin commented:
Very cool with the reconnect button. UI wise I think that the info and the reconnect button should be placed in a box at the bottom of the panel, since you hopefully only only have to use it once. 

For info about the various problems which can cause the pipe not to works, there are a few of then detected here: https://github.com/racitup/freeswitch-sounds-en-gb-richard/blob/master/tools/downsample.py

For some reason I get a Playing pop up in Audacity now. I don't remember seeing it before. There is a nice stop button on it , and when you do that, it'll go into a loop where it'll keep playing. Was something changed, which could have created this popup? Actually stopping doesn't work, when hitting stop in Blender, it keeps playing in Audacity, and I have to click the stop button on the pop up several times. 

I also noticed that an undistorted file in Blender became distorted in Audacity, so I think we may have to put the envelope points back when volume is set to 1.0. Though it illogical.

samytichadou commented:
> 
> 
> Very cool with the reconnect button. UI wise I think that the info and the reconnect button should be placed in a box at the bottom of the panel, since you hopefully only only have to use it once.

Agreed for the box, but i think it should be located on top as it is now, because the info is important to have and can stop the addon from working. Giving it doesn't take a lot of space, i think for now it's better to let it here, what do you think ?
> 
> For info about the various problems which can cause the pipe not to works, there are a few of then detected here: https://github.com/racitup/freeswitch-sounds-en-gb-richard/blob/master/tools/downsample.py

Thanks i'll take a look for future improvements !
> 
> For some reason I get a Playing pop up in Audacity now. I don't remember seeing it before. There is a nice stop button on it , and when you do that, it'll go into a loop where it'll keep playing. Was something changed, which could have created this popup?

Can you take a screenshot of it ? for me it behaves as usual, i didn't touch anything in the operators themselves, so this is weird...
> Actually stopping doesn't work, when hitting stop in Blender, it keeps playing in Audacity, and I have to click the stop button on the pop up several times.

To me, works as usual too, could it be related to the initial problem of this operators ? If the playhead goes to the end in audacity, it stops, but blender loops, so blender IS playing, audacity IS NOT playing, meaning when pressing Stop in blender, you stop blender but send `playstop:` command to audacity, and triggers play ? anyway i think it is more related to the general behavior of the play operator (giving i didn't modify the operator itself) so it could be better to discuss it [here](https://github.com/tin2tin/audacity_tools_for_blender/discussions/12) or opening an issue, what do you think ?
> 
> I also noticed that an undistorted file in Blender became distorted in Audacity, so I think we may have to put the envelope points back when volume is set to 1.0. Though it illogical.

Same here, i don't seem to have this problem... and i didn't modify the "send operators", let's open an issue ? this doesn't seem to me related to this pr specifically.

samytichadou commented:
Just put the pipe related ui elements in a box (7e3a1d8 )

![image](https://user-images.githubusercontent.com/26276003/110803823-e470fe80-827f-11eb-8264-a011807aaf9a.png)

tin2tin commented:
Just a quick note, having only a few moments to test. After setting up the path to audacity, I clicked on the pipe refresh button and audacity opened. and nothing happened. Then I closed audacity, and redid the process a couple of times, until I realized that I had to click on refresh again after audacity was opened. I haven't looked at the code, but if it isn't possible to both open than do the pipe check in one go, maybe it should first be an open audacity button and swapped with a connect button?

samytichadou commented:
> I clicked on the pipe refresh button and audacity opened. and nothing happened

that's not intended, i think it could be some problem with waiting time, can you try to set it to like s in the addon preferences ? 
And if it doesn't work, do you catch an error ? a console message ? what platform and blender version are you using ?

tin2tin commented:
I'll commit this, but we'll have to keep an eye on the delay value and try to come up with a more informative UI solution.
